src/GraphQL/Types/ArgumentInformation.cs

```diff
@@ -0,0 +1,179 @@
+        {
+            if (Expression != null)
+                return (null, Expression);
+
+            var type = TypeInformation.ConstructGraphType();
+            var argument = new QueryArgument(type)
+            {
+                Name = ParameterInfo.Name!,
+                Description = ParameterInfo.Description(),
+                DefaultValue = ParameterInfo.IsOptional ? ParameterInfo.DefaultValue : null,
+            };
+            return (argument, null);
+        }
+    }
+
+    /// <inheritdoc/>
+    public class ArgumentInformation<TReturnType> : ArgumentInformation
+    {
+        /// <inheritdoc cref="ArgumentInformation.ArgumentInformation(ParameterInfo, Type, FieldType, TypeInformation, LambdaExpression?)"/>
+        public ArgumentInformation(ParameterInfo parameterInfo, Type sourceType, FieldType fieldType, TypeInformation typeInformation, Expression<Func<IResolveFieldContext, TReturnType>>? expression)
+            : base(ValidateParameterInfo(parameterInfo), sourceType, fieldType, typeInformation, expression)
+        {
+        }
+
+        /// <inheritdoc/>
+        public ArgumentInformation(ParameterInfo parameterInfo, Type sourceType, FieldType fieldType, TypeInformation typeInformation)
+            : base(ValidateParameterInfo(parameterInfo), sourceType, fieldType, typeInformation)
+        {
+        }
+
+        /// <summary>
+        /// Validates that the <see cref="ParameterInfo"/> supplied to the constructor has a return type
+        /// that matches the <typeparamref name="TReturnType"/> of this instance.
+        /// </summary>
+        private static ParameterInfo ValidateParameterInfo(ParameterInfo parameterInfo)
+        {
+            if (parameterInfo.ParameterType != typeof(TReturnType))
+            {
+                throw new ArgumentOutOfRangeException(nameof(parameterInfo), $"Parameter must have a return type of {typeof(TReturnType).Name}.");
+            }
+            return parameterInfo;
+        }
+
+        /// <summary>
+        /// Gets or sets a delegate in the form of a <see cref="LambdaExpression"/> to be used to populate
+        /// this method argument while building the field resolver.
+        /// <br/><br/>
+        /// If not set, a query argument will be added to the field and the argument's value will be used
+        /// to populate the method argument while building the field resolver.
+        /// </summary>
+        public new Expression<Func<IResolveFieldContext, TReturnType>>? Expression
+        {
+            get => (Expression<Func<IResolveFieldContext, TReturnType>>?)base.Expression;
+            set => base.Expression = value;
+        }
+
+        /// <summary>
+        /// Applies <see cref="GraphQLAttribute"/> attributes pulled from the <see cref="ArgumentInformation.ParameterInfo">ParameterInfo</see> onto this instance.
+        /// </summary>
+        public virtual void ApplyAttributes()
```

Comment:
May be moved into base type. Actually I think entire `ArgumentInformation<TReturnType>` can be eliminated. I'll add more ideas later.

Reply:
We can experiment with that in another PR.  I already had it that way, and changed it because it is so much more useful this way.  I'd like to merge it as-is, and if you want to try some changes, that's fine.

Reply:
I'm not sure that I dare to change this code myself. Maybe. Also I can just forget about it :)

Reply:
This is sorta the point, is that it's too complex to write expression trees manually.  If you look at my "before/after" code above, you will see how much simpler it is with generics.  Unfortunately the simplification comes at a cost.  But nobody wants to write expression trees.  None of my programmers want to, anyway.  And I wouldn't want to write a sample in our docs saying, "oh yeah you can write your own attribute.... but only if you know how to create expression trees".  It's just not for the average joe.

Reply:
I added an example of what I try to tell in other thread.

Reply:
So with the design I have, all the details of working with reflection and expressions are carefully tucked away where nobody should need to deal with them.  The methods and classes that deal with that are almost all `internal`; it "just happens".  And still you can easily write a custom attribute that contains code that will get embedded directly into a precompiled field resolver for that specific field.

If you look at each of my attributes that I've written, you will see that the code is extremely simple.  Anyone can copy an attribute and change it for their needs without knowing almost anything about reflection or expressions.

Reply:
Let's continue in another conversation.